APE_Black_Formatting.rst

```diff
@@ -0,0 +1,286 @@
+
+Technical issues related to the initial transition and subsequent maintenance of
+Black formatting are surmountable. There is plenty of prior art and good
+documentation. Sociological issues related to developer preferences are
+surmountable. Evidence from other projects is that developers either embrace or
+at least accept autoformatting. Developer documentation can provide clear
+instructions to contributors on how to implement Black formatting in their
+development process. Issues with this process are no more difficult (and
+arguably easier) to resolve than enforcing PEP-8 standards.
+
+Therefore, this APE proposes to adopt Black for ``astropy``.
+
+Concerns and mitigations
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Foolish consistency
+~~~~~~~~~~~~~~~~~~~
+A concern that has been highlighted by some members of the community (including,
+initially, one of the co-authors of this APE, and highlighed in detail in `this
+blog post
+<https://luminousmen.com/post/my-unpopular-opinion-about-black-code-formatter>`_)
+is that adopting Black violates the PEP8 principle that "A Foolish Consistency
+is the Hobgoblin of Little Minds". I.e., that by using a code formatter, the
+developer thinks less about readability because they think they can let the
+formatter do this for them. While this concern has some validity, this APE
+explicitly recognizes that **developers and code reviewers are ultimately
+responsible for maintaining a readable code base**. That is, there are many
+choices concerning logic and readability of code that Black does *not* affect,
+and those choices are more important for developers to focus on than mechanistic
+decisions that Black makes about whitespace and newlines.
+
+Undesirable formatting
+~~~~~~~~~~~~~~~~~~~~~~
+On occasion Black's formatting produces undesirable results. The code may look
+very different from expectation due to mathematical or other domain-specific
+conventions, or some specific formatting that conveys meaning may be lost. A
+simple example of this is a constant identity matrix which would typically be
+written as::
+
+    matrix = [[1, 0, 0],
+              [0, 1, 0],
+              [0, 0, 1]]
+
+Black reformats this as shown below in a way that is very different from the
+typical representation for a matrix and makes it harder to read the structure::
+
+    matrix = [[1, 0, 0], [0, 1, 0], [0, 0, 1]]
+
+The mitigation here is the exception policy:  In this case the author could
+surround the matrix with the ``#fmt: off``/``#fmt: on`` escape hatch, as this
+clearly makes the code more readable.  While many examples are less clear than
+the above, and indeed in most cases "readability" is a subjective concept, this
+mechanism allows for such cases to be explicitly excepted where the developer
+thinks the improved readability warrants it.
+
+History
+~~~~~~~
+This changes almost every Python file. Git blame can skip the commit, but git
+diff won't. But we are already doing other bulk changes that hit every file.
```

Comment:
I'm not sure that touching every file is really that bad, the concern is more here that it might in fact touch most lines (which is still fine because we can mark the commit to be skipped, but I do think this is unprecedented). If you keep the last sentence, give some examples?

Reply:
Indeed this touches about a 1/3 of lines in the `time` subpackage (see below). So I'm happy to rewrite this paragraph as:
```
This changes almost every Python file and a large fraction of code lines 
(for example about one third of non-blank lines in the ``astropy.time`` 
subpackage). This is more than previous bulk changes and it is an impact that
needs to be accepted with this APE.
```